File: no14503/Main.py
N, M = map(int, input().split())
r,c,d = map(int, input().split())

robot = {'x': r, 'y': c, 'dir': d}

room = [list(map(int, input().split())) for i in range(N)]

delta = [(-1, 0), (0, 1), (1, 0), (0, -1)]

answer = 0
while True:
	if room[robot['x']][robot['y']] == 0:
		room[robot['x']][robot['y']] = 2
		answer += 1

	flag = False
	for i in range(4):
		next_dir = (robot['dir'] + 3 - i) % 4
		next_x = robot['x'] + delta[next_dir][0]
		next_y = robot['y'] + delta[next_dir][1]
		
		if room[next_x][next_y] == 0:
			robot['x'] = next_x
			robot['y'] = next_y
			robot['dir'] = next_dir
			flag = True
			break

	if flag: continue

	if not room[robot['x']-delta[robot['dir']][0]][robot['y']-delta[robot['dir']][1]] == 1:
		robot['x'] -= delta[robot['dir']][0]
		robot['y'] -= delta[robot['dir']][1]
	else:
		break

print(answer)

File: no14503/Main.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int[] NM = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		int[] rcd = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

		int[][] room = new int[NM[0]][];

		for(int i = 0; i < room.length; i++) {
			room[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		}

		int x = rcd[0];
		int y = rcd[1];
		int d = rcd[2];

		int answer = 0;

		int[][] delta = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
		whileA:
		while(true) {
			if (room[x][y] == 0) {
				room[x][y] = 2;
				answer++;
			}

			for(int i = 0; i < delta.length; i++) {
				int newD = (d+3-i) % 4;
				int newX = x + delta[newD][0];
				int newY = y + delta[newD][1];

				if (room[newX][newY] == 0) {
					x = newX;
					y = newY;
					d = newD;
					continue whileA;
				}
			}

			if (room[x-delta[d][0]][y-delta[d][1]] != 1) {
				x -= delta[d][0];
				y -= delta[d][1];
			} else {
				break;
			}
		}

		System.out.println(answer);
	}
}